File: main/balanced-parentheses/balanced-parentheses.cs
using System;
using System.Collections.Generic;

[assembly: System.Reflection.AssemblyVersion("1.0.0.0")]

internal static class Solution {
    private static bool IsBalanced(this string str)
    {
        const char start_mark = '\0'; // any char but ')', ']', '}' works here

        var closers = new Stack<char>();
        closers.Push(start_mark);

        foreach (var c in str) {
            switch (c) {
            case '(':
                closers.Push(')');
                continue;

            case '[':
                closers.Push(']');
                continue;

            case '{':
                closers.Push('}');
                continue;

            case ')':
            case ']':
            case '}':
                if (closers.Pop() != c) return false;
                continue;

            default:
                continue; // ignore non-grouping characters
            }
        }

        return closers.Pop() == start_mark; // only balanced if all are closed
    }

    private static void Main()
    {
        for (var t = int.Parse(Console.ReadLine()); t > 0; --t)
            Console.WriteLine(Console.ReadLine().IsBalanced() ? "YES" : "NO");
    }
}

File: main/balanced-parentheses/balanced_parentheses.rb
require 'set'

# balanced? method, and supporting constants, to check if groupers are balanced
class String
  GROUPERS = { '(' => ')', '[' => ']', '{' => '}' }.freeze
  CLOSERS = Set.new(GROUPERS.values).freeze

  def balanced?
    closer_stack = []

    each_char do |char|
      if (closer = GROUPERS[char])
        closer_stack.push(closer)
      elsif CLOSERS.include?(char)
        return false unless closer_stack.pop == char
      end
    end

    !closer_stack.pop # only balanced if all are closed
  end
end

gets.to_i.times { puts gets.balanced? ? 'YES' : 'NO' }
